experimental/devices/sn3218/sn3218.go

```diff
@@ -0,0 +1,142 @@
+)
+
+// Dev is a handler to sn3218 controller.
+type Dev struct {
+	i2c        i2c.Dev
+	states     [18]bool
+	brightness [18]byte
+}
+
+// New returns a handle to a SN3218 LED driver.
+func New(bus i2c.Bus) (Dev, error) {
+	d := i2c.Dev{Bus: bus, Addr: i2cAddress}
+	dev := Dev{}
+	dev.i2c = d
+	dev.reset()
+	return dev, nil
+}
+
+// Halt resets the registers and switches the driver off.
+func (d *Dev) Halt() error {
+	if err := d.Disable(); err != nil {
+		return err
+	}
+	return d.reset()
+}
+
+// Enable enables the SN3218.
+func (d *Dev) Enable() error {
+	_, err := d.i2c.Write([]byte{cmdEnableOutput, 0x01})
+	return err
+}
+
+// Disable disables the SN3218.
+func (d *Dev) Disable() error {
+	_, err := d.i2c.Write([]byte{cmdEnableOutput, 0x00})
+	return err
+}
+
+// GetState returns the state (on/off) and the brightness (0..255) of the
+// Channel 0..17.
+func (d *Dev) GetState(channel int) (bool, byte, error) {
+	if channel < 0 || channel >= 18 {
+		return false, 0, errors.New("Channel number out of range 0..17")
+	}
+	return d.states[channel], d.brightness[channel], nil
+}
+
+// Switch switched the channel (0..18) to state (on/off).
+func (d *Dev) Switch(channel int, state bool) error {
+	if channel < 0 || channel >= 18 {
+		return errors.New("Channel number out of range 0..17")
+	}
+	d.states[channel] = state
+	return d.updateStates()
+}
+
+// SetGlobalBrightness sets the brightness of all channels to the value (0..255).
+func (d *Dev) SetGlobalBrightness(value byte) {
+	for i := 0; i < 18; i++ {
+		d.brightness[i] = value
+	}
+	d.updateBrightness()
+}
+
+// SetBrightness sets the brightness of led (0..17) to value (0..255).
+func (d *Dev) SetBrightness(channel int, value byte) error {
+	if channel < 0 || channel >= 18 {
+		return errors.New("Chhannel number out of range 0..17")
+	}
+	d.brightness[channel] = value
+	d.updateBrightness()
+	return nil
+}
+
+// SwitchAll switches all channels accoring to the state (on/off).
+func (d *Dev) SwitchAll(state bool) {
+	for i := 0; i < 18; i++ {
+		d.states[i] = state
+	}
+	d.updateStates()
+}
+
+// Reset resets the registers to the default values.
+func (d *Dev) reset() error {
+	_, err := d.i2c.Write([]byte{cmdReset, 0xFF})
+	d.states = [18]bool{}
+	d.brightness = [18]byte{}
+	return err
+}
+
+func boolArrayToInt(states [18]bool) uint {
+	var result uint = 0
+	for i := uint(0); i < uint(18); i++ {
+		state := uint(1)
+		if !states[i] {
+			state = uint(0)
+		}
+		result |= (state << i)
+	}
+	return result
+}
+
+func (d *Dev) update() error {
+	_, err := d.i2c.Write([]byte{cmdUpdate, 0xFF})
+	return err
+}
+
+func (d *Dev) updateStates() error {
+	mask := boolArrayToInt(d.states)
+	_, err := d.i2c.Write([]byte{cmdEnableLeds, byte(mask & 0x3F), byte((mask >> 6) & 0x3F), byte((mask >> 12) & 0X3F)})
+	if err != nil {
+		return err
+	}
+	return d.update()
+}
+
+func (d *Dev) updateBrightness() error {
+	_, err := d.i2c.Write(append([]byte{cmdSetBrightnessValues}, d.brightness[0:len(d.brightness)]...))
```

Comment:
This causes memory allocation, instead do

```
cmd := [19]byte{cmdSetBrightnessValues}
copy(cmd[1:], d.brightness)
if _ err := d.i2c.Write(cmd[:]); err != nil {
...
```

Reply:
fixed